### PR TITLE
[KV offload][3/N] Add worker-side CPU support

### PR DESCRIPTION
**Pull request by orozery**
This PR adds worker-side support for CPU offloading.
It uses the `swap_blocks` function to perform the actual copying between CPU and GPU.
Supports any `cpu_block_size` which is divided by `gpu_block_size`.

Part of the work described in RFC #19854 
Depends on #19848

### CODE REVIEW

**Comment by orozery** [comment]
@njhill As you can see, in this PR I reused the `swap_blocks` from V0.
However, looking into the actual cuda at `cache_kernels.cu`, I see that it uses `cudaMemcpyAsync`.
This means that when the call returns, I cannot be sure that the transfer actually completed.
This violates my design assumption that my `TransferFunction` should be synchronous. 
I was thinking on modifying the current kernel to use `cudaMemcpy` instead.
This will change it for V0 as well, so maybe it's not a good idea?

The other option is to create another kernel.
I would actually prefer a kernel that handles the swap for all layers at once.
This can improve performance since it will reduce the number of python->cuda context switches by a factor of `num_layers`.
The question then is how should I expose this new kernel to the code? Should I add another API function to `AttentionBackend`?

File: vllm/v1/offloading/worker/cpu_gpu.py [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+                                        GPULoadStoreSpec)
+from vllm.v1.offloading.worker.worker import (OffloadingHandler,
+                                              TransferResult, TransferSpec)
+
+logger = init_logger(__name__)
+
+
+def block_ids(specs_list: list[LoadStoreSpec],
+              block_size_factor: int,
+              skip_count: int = 0) -> Iterator[int]:
```
Comment (ApostaC) [inline vllm/v1/offloading/worker/cpu_gpu.py]: In what case will we have `skip_count > 0`?

Based on the code, if spec_list = [0, 1, 3], block_size_factor = 4, and skip_count = 1, the output will be: "[1, 2, 3, 5, 6, 7, 13, 14, 15]", which looks a bit weird.

File: vllm/v1/offloading/worker/cpu_gpu.py [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+            dst_caches = self.cpu_caches
+            src_block_size_factor = 1
+            dst_block_size_factor = self.block_size_factor
+
+        dst_sub_blocks_to_skip = (-len(src_specs) % dst_block_size_factor)
+
+        assert (
+            len(src_specs) *
+            src_block_size_factor == len(dst_specs) * dst_block_size_factor -
+            dst_sub_blocks_to_skip)
```
Comment (ApostaC) [inline vllm/v1/offloading/worker/cpu_gpu.py]: I understand in some cases there will be a "sub block" that needs to be skipped since CPU blocks could be larger than GPU blocks, and the transfer can start in the middle of a CPU block.

However, since CPU could be both src and dst, why we only have `dst_sub_blocks_to_skip` but no "`src_sub_blocks_to_skip`"?

File: vllm/v1/offloading/worker/cpu_gpu.py [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+
+        src_to_dst_list: list[tuple[int, int]] = list(
+            zip(
+                block_ids(src_specs, src_block_size_factor),
+                block_ids(dst_specs, dst_block_size_factor,
+                          dst_sub_blocks_to_skip)))
```
Comment (ApostaC) [inline vllm/v1/offloading/worker/cpu_gpu.py]: Regarding the `skip_count`, should we only skip the sub block in the first block?

Also, will there be cases that we need to skip the end of a block? Let's say we have 10 GPU blocks, and the `block_size_factor = 4`, then we probably need to skip the last 2 blocks.

File: vllm/v1/offloading/worker/cpu_gpu.py [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+        src_to_dst = torch.tensor(src_to_dst_list,
+                                  device="cpu",
+                                  dtype=torch.int64).view(-1, 2)
+
+        event = self.events_pool.pop() if self.events_pool \
+            else torch.cuda.Event()
+        with torch.cuda.stream(stream):
+            for src_cache, dst_cache in zip(src_caches, dst_caches):
+                self.attn_backend.swap_blocks(src_cache, dst_cache, src_to_dst)
```
Comment (ApostaC) [inline vllm/v1/offloading/worker/cpu_gpu.py]: Do other attention backends also have this `swap_blocks` implemented? (such as FlashInfer and FlashMLA)

**Review by ApostaC — CHANGES_REQUESTED** [review]
Got some questions regarding the `skip_blocks` semantics and the compatibility with other attention backends.

File: vllm/v1/offloading/worker/cpu_gpu.py (hunk 1 above)
Comment (orozery) [reply vllm/v1/offloading/worker/cpu_gpu.py]: As you noted below, `skip_count > 0` is for the case CPU block size is larger than the GPU block size, AND vLLM's scheduler is requesting to load starting from a middle of a CPU block.
The `swap_blocks` function assumes block_ids are all given in GPU block size.
It is actually not aware of the CPU block size.

Assuming for example that the GPU block size is 16, and CPU block size is 64 (block_size_factor=4).
So this function (`block_ids`) translates the given CPU blocks [0,1,3] to blocks of size 16.
The first CPU block 0 matches sub-blocks [0, 1, 2, 3].
The second CPU block 1 matches sub-blocks [4, 5, 6, 7]
The third CPU block 3 matches sub-blocks [12, 13, 14, 15]
Summing up, we get the following sub-blocks:
[0, 1, 2, 3, 4, 5, 6, 7, 12, 13, 14, 15]
Now since `skip_count=1`, the first sub-block is omitted and we get:
[1, 2, 3, 4, 5, 6, 7, 12, 13, 14, 15]

File: vllm/v1/offloading/worker/cpu_gpu.py (hunk 2 above)
Comment (orozery) [reply vllm/v1/offloading/worker/cpu_gpu.py]: When storing GPU->CPU, the offloading connector always aligns writes to the CPU block size.
Thus we don't need to support `skip_count > 0` for the source specs.

File: vllm/v1/offloading/worker/cpu_gpu.py (hunk 3 above)
Comment (orozery) [reply vllm/v1/offloading/worker/cpu_gpu.py]: The only skip scenario is upto `block_size_factor` GPU blocks from the destination specs.
Skipping the last blocks is not needed since the lookup function (`get_num_new_matched_tokens`) will not lookup partial CPU blocks.

File: vllm/v1/offloading/worker/cpu_gpu.py (hunk 4 above)
Comment (orozery) [reply vllm/v1/offloading/worker/cpu_gpu.py]: In this PR I only targeted the default backend which is FlashAttention.
I think that extending support for CPU offloading to other backends, and more generally to other accelerators (TPU) should be done in follow-up PRs.

File: vllm/v1/offloading/worker/cpu_gpu.py [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+            dst_sub_blocks_to_skip)
+
+        src_to_dst_list: list[tuple[int, int]] = list(
+            zip(
+                block_ids(src_specs, src_block_size_factor),
+                block_ids(dst_specs, dst_block_size_factor,
+                          dst_sub_blocks_to_skip)))
+        src_to_dst = torch.tensor(src_to_dst_list,
+                                  device="cpu",
+                                  dtype=torch.int64).view(-1, 2)
```
Comment (njhill) [inline vllm/v1/offloading/worker/cpu_gpu.py]: I don't think the `.view` is needed?

File: vllm/v1/offloading/worker/cpu_gpu.py (hunk 4 above)
Comment (njhill) [reply vllm/v1/offloading/worker/cpu_gpu.py]: `swap_blocks` was only used in V0 and isn't implemented for the V1 attention back-ends. TBD if we want to add this to the V1 attention implementations just for this. cc @WoosukKwon 

I think it's as important to support other backends like MLA and FlashInfer ... that is the default for Blackwell.

Perhaps for now we can just call the kernel directly i.e. `_custom_ops.swap_blocks()`. However whether we need to do this for k and v separately depends on the backend. It could be determined via something like:
```pytorch
    attn_backend.get_kv_cache_shape(1, 16, 1, 1)[0] == 2
```

we could also compare performance of NIXL for this

File: vllm/v1/offloading/worker/cpu_gpu.py [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+            dst_block_size_factor = self.block_size_factor
+
+        dst_sub_blocks_to_skip = (-len(src_specs) % dst_block_size_factor)
+
+        assert (
+            len(src_specs) *
+            src_block_size_factor == len(dst_specs) * dst_block_size_factor -
+            dst_sub_blocks_to_skip)
+
+        src_to_dst_list: list[tuple[int, int]] = list(
```
Comment (njhill) [inline vllm/v1/offloading/worker/cpu_gpu.py]: Instead of constructing this list, it would probably be more efficient to allocate and populate a `num_blocks x 2` numpy array which can then be viewed as the cpu mapping tensor directly.

File: vllm/v1/offloading/worker/cpu_gpu.py [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+        # success
+        return True
+
+    def get_finished(self) -> list[TransferResult]:
+        results: list[TransferResult] = []
+        for job_id, event in list(self.transfer_events.items()):
+            if event.query():
+                results.append((job_id, True))
+                del self.transfer_events[job_id]
+                self.events_pool.append(event)
```
Comment (njhill) [inline vllm/v1/offloading/worker/cpu_gpu.py]: nit: better to avoid copying the events every time
```suggestion
        results: list[TransferResult] = []
        for job_id, event in self.transfer_events.items():
            if event.query():
                results.append((job_id, True))
                self.events_pool.append(event)
        for job_id, _ in results:
            del self.transfer_events[job_id]
```

**Review by njhill — COMMENTED** [review]
Thanks @orozery for all of the hard work on this

File: vllm/v1/offloading/worker/cpu_gpu.py (hunk 4 above)
Comment (orozery) [reply vllm/v1/offloading/worker/cpu_gpu.py]: Changed to using _custom_ops directly, so this should work with any attention backend with shape (2, num_blocks, ...)

File: vllm/v1/offloading/worker/cpu_gpu.py (hunk 6 above)
Comment (orozery) [reply vllm/v1/offloading/worker/cpu_gpu.py]: I switched to using tensors directly

File: vllm/v1/offloading/worker/cpu_gpu.py [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+        event = self.events_pool.pop() if self.events_pool \
+            else torch.cuda.Event()
+        with torch.cuda.stream(stream):
+            for src_cache, dst_cache in zip(src_caches, dst_caches):
+                src_key_cache = src_cache[0]
+                dst_key_cache = dst_cache[0]
+                ops.swap_blocks(src_key_cache, dst_key_cache, src_to_dst)
+                src_value_cache = src_cache[1]
+                dst_value_cache = dst_cache[1]
+                ops.swap_blocks(src_value_cache, dst_value_cache, src_to_dst)
```
Comment (njhill) [inline vllm/v1/offloading/worker/cpu_gpu.py]: We should handle the case block dimension is the first dimension too (includes FlashInfer). Per my earlier comment we can detect that from the attention backend and have a flag to control the behavior here.

File: vllm/v1/offloading/worker/cpu_gpu.py [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+        dst_sub_blocks_to_skip = (-src_blocks.size(0) % dst_block_size_factor)
+
+        assert (src_blocks.size(0) *
+                src_block_size_factor == dst_blocks.size(0) *
+                dst_block_size_factor - dst_sub_blocks_to_skip)
+
+        src_to_dst = torch.column_stack(
+            (expand_block_ids(src_blocks, src_block_size_factor),
+             expand_block_ids(dst_blocks,
+                              dst_block_size_factor)[dst_sub_blocks_to_skip:]))
```
Comment (njhill) [inline vllm/v1/offloading/worker/cpu_gpu.py]: This looks good!

We could potentially optimize further by creating the empty final tensor first and having the `expand_block_ids` methods write into that rather than each creating an intermediate concatenated tensor.

Comment (orozery) [reply vllm/v1/offloading/worker/cpu_gpu.py]: I implemented something like this now.

File: vllm/v1/offloading/worker/cpu_gpu.py (hunk 8 above)
Comment (orozery) [reply vllm/v1/offloading/worker/cpu_gpu.py]: Thanks!
I pushed a new implementation which I think should work for both FlashInfer and MLA (also added a test).
Also, now supporting multiple attention backends (for hybrid memory allocator).

File: vllm/v1/offloading/worker/cpu_gpu.py [hunk 10]
```diff
@@ -0,0 +1,169 @@
...
+
+    output_idx = 0
+    for block_id in block_ids:
+        base_block_id = block_id * block_size_factor
+        for i in range(skip_count, block_size_factor):
+            output[output_idx] = base_block_id + i
+            output_idx += 1
+
+        # finished skipping
+        skip_count = 0
```
Comment (njhill) [inline vllm/v1/offloading/worker/cpu_gpu.py]: updating tensor element-wise is actually very slow, I was thinking more like what you had before but passing `out` arg to `torch.cat`.

Alternatively could keep as numpy array (which I think block_ids should be already per other comment) ... which are much more efficient to manipulate.

By the way I don't think these specific optimizations should hold up getting the functionality in, can always be follow-on updates.

**Review by njhill — APPROVED** [review]
Thanks @orozery, looks great

File: vllm/v1/offloading/worker/cpu_gpu.py (hunk 10 above)
Comment (orozery) [reply vllm/v1/offloading/worker/cpu_gpu.py]: I switched to using numpy

**Review by njhill — APPROVED** [review]

File: vllm/v1/kv_offload/worker/cpu_gpu.py [hunk 11]
```diff
@@ -0,0 +1,171 @@
...
+                    src_key_cache = src_tensor[0]
+                    dst_key_cache = dst_tensor[0]
+                    ops.swap_blocks(src_key_cache, dst_key_cache,
+                                    src_to_dst_tensor)
+                    src_value_cache = src_tensor[1]
+                    dst_value_cache = dst_tensor[1]
+                    ops.swap_blocks(src_value_cache, dst_value_cache,
+                                    src_to_dst_tensor)
+                else:
+                    ops.swap_blocks(src_tensor, dst_tensor, src_to_dst_tensor)
```
Comment (njhill) [inline vllm/v1/kv_offload/worker/cpu_gpu.py:155]: Not for this PR but I think we could potentially abstract this in a follow-on via https://github.com/vllm-project/vllm/blob/087c6ffc9202599f438f1f7e0d6449020a958ac1/vllm/distributed/kv_transfer/kv_connector/v1/base.py#L143

with https://github.com/vllm-project/vllm/pull/24690

**Review by ApostaC — APPROVED** [review]
LGTM!
